File: app/src/main/java/org/butterbrot/heve/ubill/entity/Bill.java
package org.butterbrot.heve.ubill.entity;

import java.util.List;

import io.objectbox.annotation.Entity;
import io.objectbox.annotation.Id;

@Entity
public class Bill {

    @Id
    private Long id;
    private List<Fellow> fellows;
    private List<Item> items;
    private int remainder;

    public Bill(List<Fellow> fellows, List<Item> items) {
        this.fellows = fellows;
        this.items = items;
        updateRemainder();
    }

    private void updateRemainder() {
        for (Item item : items) {
            remainder += item.getRemainder();
        }
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public List<Fellow> getFellows() {
        return fellows;
    }

    public void setFellows(List<Fellow> fellows) {
        this.fellows = fellows;
    }

    public List<Item> getItems() {
        return items;
    }

    public void setItems(List<Item> items) {
        this.items = items;
    }

    public int getRemainder() {
        return remainder;
    }

    public void setRemainder(int remainder) {
        this.remainder = remainder;
    }
}

File: app/src/main/java/org/butterbrot/heve/ubill/entity/Bill.kt
package org.butterbrot.heve.ubill.entity;

import io.objectbox.annotation.Entity;
import io.objectbox.annotation.Id;
import io.objectbox.annotation.Index
import io.objectbox.relation.ToMany

@Entity
class Bill(@Index var name: String = "", @Transient private val fellowsParam: List<Fellow> = listOf(),
           @Transient private val itemsParam: List<Item> = listOf()) {

    @Id
    var id: Long = 0
    lateinit var fellows: ToMany<Fellow>
    lateinit var items: ToMany<Item>

    init {
        fellows.addAll(fellowsParam)
        items.addAll(itemsParam)
    }

    fun setFellows(fellows: List<Fellow>) {
        this.fellows.clear()
        this.fellows.addAll(fellows)
    }

    fun add(item: Item) {
        items.add(item)
    }

    override fun toString(): String {
        return name
    }

    fun  hasFellow(fellow: Fellow): Boolean {
        return fellows.any{ it == fellow }
    }

    fun fellowsInSplittings(): LongArray {
        return items.flatMap{ it.fellowsInSplittings() }.toSet().toLongArray()
    }
}